nexus/db-queries/src/db/datastore/instance.rs

```diff
@@ -1188,28 +1167,102 @@ impl DataStore {
                 );
             }
         }
-        //
-        // NOTE: from this point forward it is OK if we update the
-        // instance's `boot_disk_id` column with the updated value
-        // again. It will have already been assigned with constraint
-        // checking performed above, so updates will just be
-        // repetitive, not harmful.
 
-        let r = diesel::update(instance_dsl::instance)
+        let query = diesel::update(instance_dsl::instance)
+            .into_boxed()
             .filter(instance_dsl::id.eq(authz_instance.id()))
-            .filter(instance_dsl::state.eq_any(OK_TO_SET_BOOT_DISK_STATES))
+            .filter(instance_dsl::state.eq_any(InstanceState::NO_VMM_STATES));
+        let query = if boot_disk_id.is_some() {
+            query.filter(
+                instance_dsl::boot_disk_id
+                    .ne(boot_disk_id)
+                    .or(instance_dsl::boot_disk_id.is_null()),
+            )
+        } else {
+            query.filter(instance_dsl::boot_disk_id.is_not_null())
```

Comment:
boring: `instance_dsl::boot_disk_id.ne(boot_disk_id)` would always be `false` here since the new id is `None` aka `NULL` aka `always compares false`. i think reading this without seeing our prior conversation is much more self-explanatory, so a comment to this effect would be repetitive.. @hawkw wdyt? obvious enough in retrospect?

Reply:
fine with me!